pkg/processor/runtime/rpc/abstract.go

```diff
@@ -229,15 +230,14 @@ func (r *AbstractRuntime) SupportsControlCommunication() bool {
 }
 
 // Drain signals to the runtime to drain its accumulated events and waits for it to finish
-func (r *AbstractRuntime) Drain() error {
+func (r *AbstractRuntime) Drain(drainDoneControlMessageChan chan *controlcommunication.ControlMessage) error {
 	// we use SIGUSR2 to signal the wrapper process to drain events
 	if err := r.signal(syscall.SIGUSR2); err != nil {
 		return errors.Wrap(err, "Failed to signal wrapper process to drain")
 	}
 
-	// wait for process to finish event handling or timeout
-	// TODO: replace the following function with one that waits for a control communication message or timeout
-	r.waitForProcessTermination(r.configuration.WorkerTerminationTimeout)
+	// wait for the process to be finished event handling or timeout
+	r.waitForDrainingFinish(drainDoneControlMessageChan, r.configuration.WorkerTerminationTimeout)
 
 	return nil
 }
```

Comment:
I think we have 2 options:
1. Create a new channel, subscribe and pass to `waitForDraining..` on each call to Drain.
2. Create one channel and subscribe it on runtime start (or `startWrapper`), and use `r.drainDoneChannel` in the wait function. Then unsubscribe on runtime's `Stop`.

IMO - 2 is cleaner.

Reply:
@TomerShor I think that 1st is better, because writing to channel is blocking operation, so this approach guarantees fastest delivery of messages since we can Drain the same worker from many partitions and as a result block of read/write operations

Reply:
So create a new channel for each partition invoking `Drain`, because we send the drainDone message anyways?

Reply:
@TomerShor didn't get it